numpy/_core/src/multiarray/ctors.c

```diff
@@ -720,7 +720,46 @@ PyArray_NewFromDescr_int(
         }
     }
 
-    fa = (PyArrayObject_fields *) subtype->tp_alloc(subtype, 0);
+    npy_bool data_on_instance = NPY_FALSE;
+    if (subtype == &PyArray_Type) {
```

Comment:
In principle one could add a check for `tp_basicsize == our size` and `tp_alloc == PyObject_GenericAlloc` (with alignment logic without the tp_basicsize check).  But in the end, this is mainly about scalars so subclasses do not matter.